In [1]:
import sqlalchemy
%load_ext sql
%sql oracle://scott:tiger@localhost:1521/orcl

# 여러 테이블의 데이터를 조인해서 출력하기 (EQUI JOIN)
"조인을 이용하면 두 개 이상의 테이블들의 컬럼들을 하나의 결과로 모아서 볼수 가 있다."  
ex) 이름, 부서위치를 출력하시오

In [2]:
%%sql

select ename, loc
    from emp, dept

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('jack', 'NEW YORK'),
 ('KING', 'NEW YORK'),
 ('BLAKE', 'NEW YORK'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'NEW YORK'),
 ('MARTIN', 'NEW YORK'),
 ('ALLEN', 'NEW YORK'),
 ('TURNER', 'NEW YORK'),
 ('JAMES', 'NEW YORK'),
 ('WARD', 'NEW YORK'),
 ('FORD', 'NEW YORK'),
 ('SMITH', 'NEW YORK'),
 ('SCOTT', 'NEW YORK'),
 ('ADAMS', 'NEW YORK'),
 ('MILLER', 'NEW YORK'),
 ('jack', 'DALLAS'),
 ('KING', 'DALLAS'),
 ('BLAKE', 'DALLAS'),
 ('CLARK', 'DALLAS'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'DALLAS'),
 ('ALLEN', 'DALLAS'),
 ('TURNER', 'DALLAS'),
 ('JAMES', 'DALLAS'),
 ('WARD', 'DALLAS'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'DALLAS'),
 ('jack', 'CHICAGO'),
 ('KING', 'CHICAGO'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'CHICAGO'),
 ('JONES', 'CHICAGO'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'CHICAGO'),
 ('SMITH', 'CHICAGO'),
 ('SCOTT', 'CHICAGO'),
 ('ADAMS', 'CHICAGO'),
 ('MILLER', 'CHICAGO'),
 ('jack', 'BOSTON'),
 ('KING', 'BOSTON'),
 ('BLAKE', 'BOSTON'),
 ('CLARK', 'BOSTON'),
 ('JONES', 'BOSTON'),
 ('MARTIN', 'BOSTON'),
 ('ALLEN', 'BOSTON'),
 ('TURNER', 'BOSTON'),
 ('JAMES', 'BOSTON'),
 ('WARD', 'BOSTON'),
 ('FORD', 'BOSTON'),
 ('SMITH', 'BOSTON'),
 ('SCOTT', 'BOSTON'),
 ('ADAMS', 'BOSTON'),
 ('MILLER', 'BOSTON')]

14 x 4건 =56건이 되어서 다 조인되었음  
king이 new york에도 근무하고 dallas, chicago, boston에도 근무하는 꼴

In [3]:
%%sql

select ename, loc
    from emp, dept
    where emp.deptno = dept.deptno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'NEW YORK')]

조인 조건을 where 절에 기술해야 정확한 결과를 볼수 있음  
조인 조건을 줘야 조인을 할 수 있다.  
emp 테이블의 deptno는 dept 테이블에 deptno와 같다라고 오라클에게 알려주면서  
이 컬럼으로 두 개의 테이블을 서로 조인(join)하는 것

<br>

문제 210. 직업이 SALESMAN인 사원들의 이름과 월급과 직업과 부서위치를 출력하시오

In [4]:
%%sql

select ename, sal, job, loc
    from emp e, dept d
    where e.deptno = d.deptno 
          and e.job = 'SALESMAN'

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('MARTIN', Decimal('1250'), 'SALESMAN', 'CHICAGO'),
 ('ALLEN', Decimal('1600'), 'SALESMAN', 'CHICAGO'),
 ('TURNER', Decimal('1500'), 'SALESMAN', 'CHICAGO'),
 ('WARD', Decimal('1250'), 'SALESMAN', 'CHICAGO')]

문제 211. 월급이 2000 이상인 사원들의 이름과 월급과 부서위치를 출력하시오

In [5]:
%%sql

select ename, sal, loc
    from emp e, dept d
    where e.deptno = d.deptno
          and sal >= 2000

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', Decimal('5000'), 'NEW YORK'),
 ('BLAKE', Decimal('2850'), 'CHICAGO'),
 ('CLARK', Decimal('2450'), 'NEW YORK'),
 ('JONES', Decimal('2975'), 'DALLAS'),
 ('FORD', Decimal('3000'), 'DALLAS'),
 ('SCOTT', Decimal('3000'), 'DALLAS')]

문제 212. 위의 결과에서 이름, 월급, 부서위치 옆에 부서번호도 같이 출력하시오

In [6]:
%%sql

select ename, sal, loc, d.deptno
    from emp e, dept d
    where e.deptno = d.deptno
          and sal >= 2000

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', Decimal('5000'), 'NEW YORK', 10),
 ('BLAKE', Decimal('2850'), 'CHICAGO', 30),
 ('CLARK', Decimal('2450'), 'NEW YORK', 10),
 ('JONES', Decimal('2975'), 'DALLAS', 20),
 ('FORD', Decimal('3000'), 'DALLAS', 20),
 ('SCOTT', Decimal('3000'), 'DALLAS', 20)]

첫째열을 select ename, sal, loc, deptno로 쓰면  
"1행의 오류: 열의 정의가 애매합니다"  
라는 오류가 뜬다  
이는 emp, dept 두 테이블 모두 deptno 컬럼이 존재하기 때문에 어느걸 출력해야 할지 애매하기 때문  

<br>

- 에러가 나지 않는다고 컬럼명 앞에 테이블 명을 적지 않는 습관을 지양하고 무조건 조인문장에서는 컬럼명 앞에 테이블 명을 적어준다.
    - 테이블.컬럼명으로 작성해줘야 검색 속도가 더 빨라짐
    - 테이블명을 붙여주지 않으면 오라클이 일일히 검색하기 때문
    
문제 213. 월급이 1000에서 3000 사이인 사원들의 이름과 월급과 부서위치를 출력하시오

In [7]:
%%sql

select e.ename, e.sal, d.loc
    from emp e, dept d
    where e.deptno = d.deptno
          and e.sal between 1000 and 3000

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('BLAKE', Decimal('2850'), 'CHICAGO'),
 ('CLARK', Decimal('2450'), 'NEW YORK'),
 ('JONES', Decimal('2975'), 'DALLAS'),
 ('MARTIN', Decimal('1250'), 'CHICAGO'),
 ('ALLEN', Decimal('1600'), 'CHICAGO'),
 ('TURNER', Decimal('1500'), 'CHICAGO'),
 ('WARD', Decimal('1250'), 'CHICAGO'),
 ('FORD', Decimal('3000'), 'DALLAS'),
 ('SCOTT', Decimal('3000'), 'DALLAS'),
 ('ADAMS', Decimal('1100'), 'DALLAS'),
 ('MILLER', Decimal('1300'), 'NEW YORK')]

문제 214. 사원번호가 7788, 7902, 7369번인 사원의 사원번호, 이름, 월급, 부서위치를 출력하시오

In [8]:
%%sql

select e.empno, e.ename, e.sal, d.loc
    from emp e, dept d
    where e.deptno = d.deptno
          and e.empno in (7788, 7902, 7369)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(7902, 'FORD', Decimal('3000'), 'DALLAS'),
 (7369, 'SMITH', Decimal('800'), 'DALLAS'),
 (7788, 'SCOTT', Decimal('3000'), 'DALLAS')]

문제 215. 이름의 첫번째 철자가 S로 시작하는 사원의 이름과 월급과 부서위치를 출력하시오

In [9]:
%%sql

select e.ename, e.sal, d.loc
    from emp e, dept d
    where d.deptno = e.deptno
          and e.ename like 'S%'

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SMITH', Decimal('800'), 'DALLAS'), ('SCOTT', Decimal('3000'), 'DALLAS')]

문제 216. DALLAS 에서 근무하는 사원들의 이름, 직업, 부서위치를 출력하시오.

In [10]:
%%sql

select e.ename, e.job, d.loc
    from emp e, dept d
    where e.deptno = d.deptno
          and d.loc = 'DALLAS'

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('JONES', 'MANAGER', 'DALLAS'),
 ('FORD', 'ANALYST', 'DALLAS'),
 ('SMITH', 'CLERK', 'DALLAS'),
 ('SCOTT', 'ANALYST', 'DALLAS'),
 ('ADAMS', 'CLERK', 'DALLAS')]

문제 217. 부서위치, 부서위치별 토탈월급을 출력하시오

In [11]:
%%sql

select d.loc, sum(e.sal)
    from emp e, dept d
    where e.deptno = d.deptno
    group by d.loc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 8750), ('DALLAS', 10875), ('CHICAGO', 9400)]

문제 218. 부서위치, 부서위치별 평균월급을 출력하는데 소수점 이하는 안나오게 반올림하고  
             부서위치별 평균월급이 높은 것 부터 출력하시오

In [12]:
%%sql

select d.loc, round(avg(e.sal), 0) as "avg(sal)"
    from dept d, emp e
    where d.deptno = e.deptno
    group by d.loc
    order by avg(sal) desc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 2917), ('DALLAS', 2175), ('CHICAGO', 1567)]

문제 219. 월급이 2700 이상인 이상인 사원들의 이름, 월급, 부서위치를 출력하시오

In [13]:
%%sql

select e.ename, e.sal, d.loc
    from emp e, dept d
    where e.deptno = d.deptno
          and e.sal >= 2700

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', Decimal('5000'), 'NEW YORK'),
 ('BLAKE', Decimal('2850'), 'CHICAGO'),
 ('JONES', Decimal('2975'), 'DALLAS'),
 ('FORD', Decimal('3000'), 'DALLAS'),
 ('SCOTT', Decimal('3000'), 'DALLAS')]

※ 지금은 테이블 명이 작아서 괜찮지만 현업에서는 테이블 명의 규격이 있다.  
   테이블 별칭을 사용해서 SQL 코딩을 좀 심플하게 작성  
   반드시 조인문장 작성할 때, 컬럼명 앞에 테이블 별칭을 사용해야 함  


※ 테이블 별칭을 사용하면 좋은 이점  

1. 검색속도가 빨라짐  
2. SQL유지보수가 쉬워짐. (가독성)

문제 220. 이름의 끝글자가 T로 끝나는 사원들의 이름, 월급, 부서위치, 부서명을 출력하시오

In [14]:
%%sql

select e.ename, e.sal, d.loc, d.dname
    from emp e, dept d
    where e.deptno = d.deptno
          and e.ename like '%T'

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SCOTT', Decimal('3000'), 'DALLAS', 'RESEARCH')]

문제 221. 직업이 SALESMAN이고 월급이 1200 이상인 사원들의 이름, 직업, 부서위치, 월급을 출력하시오

In [15]:
%%sql

select e.ename, e.job, d.loc, e.sal
    from emp e, dept d
    where e.deptno = d.deptno
          and e.job = 'SALESMAN'
          and e.sal>= 1200

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('MARTIN', 'SALESMAN', 'CHICAGO', Decimal('1250')),
 ('ALLEN', 'SALESMAN', 'CHICAGO', Decimal('1600')),
 ('TURNER', 'SALESMAN', 'CHICAGO', Decimal('1500')),
 ('WARD', 'SALESMAN', 'CHICAGO', Decimal('1250'))]

문제 222. 부서위치, 부서위치별 토탈월급을 출력하는데 DALLAS는 제외하고 출력하시오

In [16]:
%%sql

select case grouping(d.loc) when 1 then 'Total' else d.loc end as "loc", 
       sum(e.sal) as "sal"
    from dept d, emp e
    where d.deptno = e.deptno
          and d.loc != 'DALLAS'
    group by rollup(d.loc)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('CHICAGO', 9400), ('NEW YORK', 8750), ('Total', 18150)]

문제 223. 지금 출력된 결과를 다시 출력하는데 토탈월급이 높은 것 부터 출력하시오

In [17]:
%%sql

select case grouping(d.loc) when 1 then 'Total' else d.loc end as "loc", 
       sum(e.sal) as "sal"
    from dept d, emp e
    where d.deptno = e.deptno
          and d.loc != 'DALLAS'
    group by rollup(d.loc)
    order by sum(e.sal) desc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('Total', 18150), ('CHICAGO', 9400), ('NEW YORK', 8750)]

# 여러 테이블의 데이터를 조인해서 출력하기 2(NON EQUI JOIN)
"조인하려는 두개의 테이블 사이에 공통된 컬럼이 없었을 때 사용하는 조인문법"  

Ex) 이름, 월급 grade(급여등급)을 출력하시오!

In [18]:
%%sql

select e.ename, e.sal, s.grade
            from emp e, salgrade s
            where e.sal between s.losal and s.hisal

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SMITH', Decimal('800'), 1),
 ('JAMES', Decimal('950'), 1),
 ('ADAMS', Decimal('1100'), 1),
 ('WARD', Decimal('1250'), 2),
 ('MARTIN', Decimal('1250'), 2),
 ('MILLER', Decimal('1300'), 2),
 ('TURNER', Decimal('1500'), 3),
 ('ALLEN', Decimal('1600'), 3),
 ('CLARK', Decimal('2450'), 4),
 ('BLAKE', Decimal('2850'), 4),
 ('JONES', Decimal('2975'), 4),
 ('FORD', Decimal('3000'), 4),
 ('SCOTT', Decimal('3000'), 4),
 ('jack', Decimal('3500'), 5),
 ('KING', Decimal('5000'), 5)]

문제 224. 위의 결과에서 등급이 3등급인 사원들만 출력하시오

In [19]:
%%sql

select e.ename, e.sal, s.grade
    from emp e, salgrade s
    where e.sal between s.losal and s.hisal
          and s.grade = 3

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('ALLEN', Decimal('1600'), 3), ('TURNER', Decimal('1500'), 3)]

![image](https://user-images.githubusercontent.com/70187490/129708992-18c7c0f6-a68a-42db-ab73-758bbcb70911.png)

문제 225. 급여등급(grade), 급여등급별로 해당하는 사원들의 이름을 가로로 출력하시오

In [20]:
%%sql

select s.grade, listagg(e.ename, ', ') within group (order by e.ename) 이름
    from salgrade s, emp e
    where e.sal between s.losal and s.hisal
    group by s.grade

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(1, 'ADAMS, JAMES, SMITH'),
 (2, 'MARTIN, MILLER, WARD'),
 (3, 'ALLEN, TURNER'),
 (4, 'BLAKE, CLARK, FORD, JONES, SCOTT'),
 (5, 'KING, jack')]

문제 226. 급여등급(grade), 급여등급별로 해당하는 사원들의 이름을 가로로 출력하고 월급도 옆에 같이 나오게 하시오!

In [21]:
%%sql

select s.grade, listagg(e.ename||'('||e.sal||')',', ') within group (order by e.ename) 이름
    from salgrade s, emp e
    where e.sal between s.losal and s.hisal
    group by s.grade

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(1, 'ADAMS(1100), JAMES(950), SMITH(800)'),
 (2, 'MARTIN(1250), MILLER(1300), WARD(1250)'),
 (3, 'ALLEN(1600), TURNER(1500)'),
 (4, 'BLAKE(2850), CLARK(2450), FORD(3000), JONES(2975), SCOTT(3000)'),
 (5, 'KING(5000), jack(3500)')]

# 여러 테이블의 데이터를 조인해서 출력하기 3(OUTER JOIN)
"조인하려는 두 테이블의 공통된 컬럼인 deptno의 데이터가 서로 일치하지 않을 때 조인하기 위해서 사용하는 조인방법"  
ex) 사원 테이블에서 부서번호를 출력하는데 중복제거해서 출력하시오

In [22]:
%%sql

select distinct(deptno)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(70,), (30,), (10,), (20,)]

Ex) 부서(dept) 테이블에서 부서번호를 출력하시오

In [23]:
%%sql

select deptno
    from dept

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(10,), (20,), (30,), (40,)]

문제 227. emp와 dept를 서로 조인해서 이름, 부서위치, 부서번호를 출력하시오

In [24]:
%%sql

select e.ename, d.loc, d.deptno
    from emp e, dept d
    where e.deptno = d.deptno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK', 10),
 ('BLAKE', 'CHICAGO', 30),
 ('CLARK', 'NEW YORK', 10),
 ('JONES', 'DALLAS', 20),
 ('MARTIN', 'CHICAGO', 30),
 ('ALLEN', 'CHICAGO', 30),
 ('TURNER', 'CHICAGO', 30),
 ('JAMES', 'CHICAGO', 30),
 ('WARD', 'CHICAGO', 30),
 ('FORD', 'DALLAS', 20),
 ('SMITH', 'DALLAS', 20),
 ('SCOTT', 'DALLAS', 20),
 ('ADAMS', 'DALLAS', 20),
 ('MILLER', 'NEW YORK', 10)]

위의 결과를 보면 dept 테이블의 40번이 출력되지 않음을 확인할 수 있다.  
이유는 사원 테이블에 40번 없기 때문이다.  
이를 출력하기 위해선 outer join 기법을 사용해야 한다.

In [25]:
%%sql

select e.ename, d.loc, d.deptno
    from emp e, dept d
    where e.deptno(+) = d.deptno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK', 10),
 ('BLAKE', 'CHICAGO', 30),
 ('CLARK', 'NEW YORK', 10),
 ('JONES', 'DALLAS', 20),
 ('MARTIN', 'CHICAGO', 30),
 ('ALLEN', 'CHICAGO', 30),
 ('TURNER', 'CHICAGO', 30),
 ('JAMES', 'CHICAGO', 30),
 ('WARD', 'CHICAGO', 30),
 ('FORD', 'DALLAS', 20),
 ('SMITH', 'DALLAS', 20),
 ('SCOTT', 'DALLAS', 20),
 ('ADAMS', 'DALLAS', 20),
 ('MILLER', 'NEW YORK', 10),
 (None, 'BOSTON', 40)]

※   (+): Outer join sign
- outer join sign은 결과로 출력될 때 데이터가 모자란 쪽에 붙여준다.
- emp 테이블에서는 부서번호가 10, 20, 30이 었고 40번이 없다.
- dept 테이블에는 부서번호가 10, 20, 30, 40이 있다.

문제 228. 부서위치, 부서위치별 토탈월급을 출력하시오

In [26]:
%%sql

select d.loc, decode(sum(e.sal), null, 0, sum(e.sal)) as "Total sal"
    from dept d, emp e
    where e.deptno(+) = d.deptno 
    group by d.loc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 8750), ('DALLAS', 10875), ('CHICAGO', 9400), ('BOSTON', 0)]

문제 230. 우리반 테이블과 telecom_price 테이블을 조인해서  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
이름, 나이, 성별, 통신사, 통신사 기본요금(price)을 출력하는데  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;
나이가 27 이상인 학생들만 출력되게 하시오

In [27]:
%%sql

select replace(e.ename, 
               substr(e.ename, 
                              case when length(e.ename)=4 then 3 else 2 end, 1)
                                                                               ,'*') as name,
       substr(e.age, 1, 1)||'*' as age,
       e.gender, 
       e.telecom,
       t.price
        
    from telecom_price t, emp12 e
    where t.telecom = e.telecom 
          and e.age >= 27

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('남궁*미', '2*', '여', 'kt', 17000),
 ('이*혁', '2*', '남', 'kt', 17000),
 ('한*', '3*', '남', 'kt', 17000),
 ('현*연', '3*', '여', 'sk', 18500),
 ('성*창', '2*', '남', 'sk', 18500),
 ('유*영', '2*', '여', 'sk', 18500),
 ('김*민', '2*', '남', 'lg', 18000),
 ('김*원', '2*', '여', 'kt', 17000),
 ('김*승', '2*', '여', 'lg', 18000),
 ('김*린', '2*', '여', 'kt', 17000),
 ('신*종', '2*', '남', 'lg', 18000),
 ('김*순', '3*', '남', 'sk', 18500),
 ('허*민', '3*', '여', 'sk', 18500),
 ('이*성', '2*', '남', 'sk', 18500),
 ('황*현', '2*', '여', 'kt', 17000),
 ('권*환', '2*', '남', 'kt', 17000),
 ('송*미', '2*', '여', 'kt', 17000),
 ('김*라', '2*', '여', 'kt', 17000),
 ('권*원', '3*', '남', 'sk', 18500),
 ('김*원', '4*', '여', 'lg', 18000)]

# 여러 테이블의 데이터를 조인해서 출력하기 4(SELF JOIN)
"자기 자신의 테이블과 조인하는 조인 문법"  
※ 자기 자신의 테이블과 조인을 하는 이유  
    -> 사원 테이블에서 사원이름과 그 사원을 관리하는 관리자의 이름을 하나의 결과로 볼 수 있기 때문  

<br>

Ex) 사원번호, 사원이름, 관리자 번호(mgr)을 출력하시오!

In [28]:
%%sql

select empno, ename, mgr
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(122, 'jack', None),
 (7839, 'KING', None),
 (7698, 'BLAKE', 7839),
 (7782, 'CLARK', 7839),
 (7566, 'JONES', 7839),
 (7654, 'MARTIN', 7698),
 (7499, 'ALLEN', 7698),
 (7844, 'TURNER', 7698),
 (7900, 'JAMES', 7698),
 (7521, 'WARD', 7698),
 (7902, 'FORD', 7566),
 (7369, 'SMITH', 7902),
 (7788, 'SCOTT', 7566),
 (7876, 'ADAMS', 7788),
 (7934, 'MILLER', 7782)]

7839 KING  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;▲  
7566 JONES  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;▲&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;-> 한명의 mgr은 다른 한명의 empno가 될 수 있다.  
7902 FORD&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;이는 emp 테이블의 관리구조를 알수 있게 하는 데이터  
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;▲  
7369 SMITH  

<br>

예제 사원이름, 해당 사원의 관리자의 이름을 출력하시오


In [29]:
%%sql

select 사원.ename, 관리자.ename
    from emp 사원, emp 관리자
    where 사원.mgr = 관리자.empno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('BLAKE', 'KING'),
 ('CLARK', 'KING'),
 ('JONES', 'KING'),
 ('MARTIN', 'BLAKE'),
 ('ALLEN', 'BLAKE'),
 ('TURNER', 'BLAKE'),
 ('JAMES', 'BLAKE'),
 ('WARD', 'BLAKE'),
 ('MILLER', 'CLARK'),
 ('FORD', 'JONES'),
 ('SCOTT', 'JONES'),
 ('SMITH', 'FORD'),
 ('ADAMS', 'SCOTT')]

문제 231. 위의 결과를 다시 출력하는데 컬럼명을 한글로 사원 관리자라고 출력되게 하시오

In [30]:
%%sql

select 사원.ename 사원, 관리자.ename 관리자
    from emp 사원, emp 관리자
    where 사원.mgr = 관리자.empno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('BLAKE', 'KING'),
 ('CLARK', 'KING'),
 ('JONES', 'KING'),
 ('MARTIN', 'BLAKE'),
 ('ALLEN', 'BLAKE'),
 ('TURNER', 'BLAKE'),
 ('JAMES', 'BLAKE'),
 ('WARD', 'BLAKE'),
 ('MILLER', 'CLARK'),
 ('FORD', 'JONES'),
 ('SCOTT', 'JONES'),
 ('SMITH', 'FORD'),
 ('ADAMS', 'SCOTT')]

문제 232. 사원이름, 사원월급, 관리자 이름, 관리자의 월급을 출력하시오

In [31]:
%%sql

select emp.ename "사원 이름", emp.sal "사원 월급", mgr.ename "관리자 이름", mgr.sal "관리자 월급"
    from emp emp, emp mgr
    where emp.mgr = mgr.empno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('BLAKE', Decimal('2850'), 'KING', Decimal('5000')),
 ('CLARK', Decimal('2450'), 'KING', Decimal('5000')),
 ('JONES', Decimal('2975'), 'KING', Decimal('5000')),
 ('MARTIN', Decimal('1250'), 'BLAKE', Decimal('2850')),
 ('ALLEN', Decimal('1600'), 'BLAKE', Decimal('2850')),
 ('TURNER', Decimal('1500'), 'BLAKE', Decimal('2850')),
 ('JAMES', Decimal('950'), 'BLAKE', Decimal('2850')),
 ('WARD', Decimal('1250'), 'BLAKE', Decimal('2850')),
 ('MILLER', Decimal('1300'), 'CLARK', Decimal('2450')),
 ('FORD', Decimal('3000'), 'JONES', Decimal('2975')),
 ('SCOTT', Decimal('3000'), 'JONES', Decimal('2975')),
 ('SMITH', Decimal('800'), 'FORD', Decimal('3000')),
 ('ADAMS', Decimal('1100'), 'SCOTT', Decimal('3000'))]

문제 233. 위의 결과를 다시 출력하는데 사원의 월급이 관리자의 월급보다 더 큰 사원들만 출력하시오!

In [32]:
%%sql

select emp.ename "사원 이름", emp.sal "사원 월급", mgr.ename "관리자 이름", mgr.sal "관리자 월급"
    from emp emp, emp mgr
    where emp.mgr = mgr.empno
          and emp.sal >= mgr.sal

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('FORD', Decimal('3000'), 'JONES', Decimal('2975')),
 ('SCOTT', Decimal('3000'), 'JONES', Decimal('2975'))]

문제 234. 관리자보다 먼저 입사한 사원들의 사원 이름과 사원 입사일, 
             관리자 이름과 관리자 입사일을 출력하시오

In [33]:
%%sql

select 사원.ename "사원 이름", to_char(사원.hiredate, 'YYYY-MM-DD') "사원 입사일", 관리자.ename "관라지 이름", to_char(관리자.hiredate, 'YYYY-MM-DD') "관리자 입사일"
    from emp 사원, emp 관리자
    where 사원.mgr = 관리자.empno
          and 사원.hiredate < 관리자.hiredate

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('BLAKE', '1981-05-01', 'KING', '1981-11-17'),
 ('CLARK', '1981-05-09', 'KING', '1981-11-17'),
 ('JONES', '1981-04-01', 'KING', '1981-11-17'),
 ('ALLEN', '1981-02-11', 'BLAKE', '1981-05-01'),
 ('WARD', '1981-02-23', 'BLAKE', '1981-05-01'),
 ('SMITH', '1980-12-09', 'FORD', '1981-12-11')]

문제 235. 관리자의 이름을 출력하고 
             그 옆에 해당 관리자에 속한 사원들의 이름을 가로로 출력하시오


In [34]:
%%sql

select 관리자.ename "관리자", listagg(사원.ename, ' ,') within group(order by 사원.ename asc) "사원"
    from emp 관리자, emp 사원
    where 관리자.empno = 사원.mgr
    group by 관리자.ename

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('BLAKE', 'ALLEN ,JAMES ,MARTIN ,TURNER ,WARD'),
 ('CLARK', 'MILLER'),
 ('FORD', 'SMITH'),
 ('JONES', 'FORD ,SCOTT'),
 ('KING', 'BLAKE ,CLARK ,JONES'),
 ('SCOTT', 'ADAMS')]

문제 236. 직업, 이름, 월급, 순위를 출력하는데 순위가 월급이 높은 순서대로 순위를 부여하시오

In [35]:
%%sql

select job, ename, sal, rank () over (order by sal desc)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('PRESIDENT', 'KING', Decimal('5000'), 1),
 ('SALESMAN', 'jack', Decimal('3500'), 2),
 ('ANALYST', 'SCOTT', Decimal('3000'), 3),
 ('ANALYST', 'FORD', Decimal('3000'), 3),
 ('MANAGER', 'JONES', Decimal('2975'), 5),
 ('MANAGER', 'BLAKE', Decimal('2850'), 6),
 ('MANAGER', 'CLARK', Decimal('2450'), 7),
 ('SALESMAN', 'ALLEN', Decimal('1600'), 8),
 ('SALESMAN', 'TURNER', Decimal('1500'), 9),
 ('CLERK', 'MILLER', Decimal('1300'), 10),
 ('SALESMAN', 'WARD', Decimal('1250'), 11),
 ('SALESMAN', 'MARTIN', Decimal('1250'), 11),
 ('CLERK', 'ADAMS', Decimal('1100'), 13),
 ('CLERK', 'JAMES', Decimal('950'), 14),
 ('CLERK', 'SMITH', Decimal('800'), 15)]

문제 238. 부서위치, 이름, 월급, 순위를 출력하는데 
             부서위치별로 각각 윌급이 높은 순서대로 순위를 부여하시오

In [36]:
%%sql

select d.loc, e.ename, e.sal, rank () over (partition by d.loc
                                            order by e.sal desc)
    from emp e, dept d
    where d.deptno = e.deptno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('CHICAGO', 'BLAKE', Decimal('2850'), 1),
 ('CHICAGO', 'ALLEN', Decimal('1600'), 2),
 ('CHICAGO', 'TURNER', Decimal('1500'), 3),
 ('CHICAGO', 'WARD', Decimal('1250'), 4),
 ('CHICAGO', 'MARTIN', Decimal('1250'), 4),
 ('CHICAGO', 'JAMES', Decimal('950'), 6),
 ('DALLAS', 'FORD', Decimal('3000'), 1),
 ('DALLAS', 'SCOTT', Decimal('3000'), 1),
 ('DALLAS', 'JONES', Decimal('2975'), 3),
 ('DALLAS', 'ADAMS', Decimal('1100'), 4),
 ('DALLAS', 'SMITH', Decimal('800'), 5),
 ('NEW YORK', 'KING', Decimal('5000'), 1),
 ('NEW YORK', 'CLARK', Decimal('2450'), 2),
 ('NEW YORK', 'MILLER', Decimal('1300'), 3)]


문제 239. 부서위치, 부서위치별로 속한 사원들의 이름을 가로로 출력하시오

In [37]:
%%sql

select d.loc, listagg(e.ename, ', ')
    from emp e, dept d
    where e.deptno = d.deptno
    group by d.loc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 'KING, CLARK, MILLER'),
 ('DALLAS', 'JONES, FORD, SMITH, SCOTT, ADAMS'),
 ('CHICAGO', 'BLAKE, MARTIN, ALLEN, TURNER, JAMES, WARD')]

문제 240. 부서위치, 부서위치별 토탈월급을 출력하는데
             맨 아래에 전체 토탈월급이 출력되게 하시오

In [38]:
%%sql

select case when grouping(d.loc)=0 then d.loc else 'Total Sal' end as "loc", sum(e.sal)
    from dept d, emp e
    where d.deptno = e.deptno
    group by rollup(d.loc)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('CHICAGO', 9400),
 ('DALLAS', 10875),
 ('NEW YORK', 8750),
 ('Total Sal', 29025)]

In [39]:
%%sql

select d.loc, sum(e.sal)
    from emp e, dept d
    where e.deptno = d.deptno
    group by grouping sets(d.loc,())

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('CHICAGO', 9400), ('DALLAS', 10875), ('NEW YORK', 8750), (None, 29025)]

문제 241. 직업, 직업별 최대월급, 직업별 최소월급, 직업별 평균월급을 출력하시오

In [40]:
%%sql

select job, max(sal), min(sal), avg(sal)
    from emp
    group by job

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SALESMAN', 3500, 1250, 1820),
 ('CLERK', 1300, 800, Decimal('1037.5')),
 ('ANALYST', 3000, 3000, 3000),
 ('MANAGER', 2975, 2450, Decimal('2758.333333333333333333333333333333333333')),
 ('PRESIDENT', 5000, 5000, 5000)]

문제 242. 부서위치, 부서위치별 최대월급, 부서위치별 최소월급, 부서위치별 인원수를 출력하시오

In [41]:
%%sql

select d.loc, max(e.sal), min(e.sal), count(e.ename)
    from dept d, emp e
    where d.deptno = e.deptno
    group by d.loc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 5000, 1300, 3),
 ('DALLAS', 3000, 800, 5),
 ('CHICAGO', 2850, 950, 6)]

# 여러 테이블의 데이터를 조인해서 출력하기 5(ON절)
1999 ANSI 조인문법의 시작  
  
- 조인문법 2가지  
    1. 오라클 조인 문법
        - Equi join
        - Non equi join
        - Self join
        - Outer join
    2. 1999 ANSI 조인 문법
         - On join
         - Using join
         - Natural join
         - Left / Right / Full outer join
         - Cross join


문제 243. 이름과 월급과 부서위치를 출력하는데 월급이 2400 이상인 사원들만 출력하시오
             (on절을 사용한 join 문법으로 수행)

In [42]:
%%sql

select e.ename, e.sal, d.loc
    from emp e join dept d
    on (e.deptno = d.deptno)
    where e.sal >= 2400

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', Decimal('5000'), 'NEW YORK'),
 ('BLAKE', Decimal('2850'), 'CHICAGO'),
 ('CLARK', Decimal('2450'), 'NEW YORK'),
 ('JONES', Decimal('2975'), 'DALLAS'),
 ('FORD', Decimal('3000'), 'DALLAS'),
 ('SCOTT', Decimal('3000'), 'DALLAS')]

※ on 절을 사용한 조인문법은 조인 조건은 on 절에 주게 되어 있고  
검색 조건은 where 절에 주게끔 구분해 놓았음

문제 244. DALLAS 에서 근무하는 사원들의 이름과 월급과 부서위치와 직업을 출력하는데
             on절을 사용한 join 문법으로 수행하시오

In [43]:
%%sql

select e.ename, e.sal, d.loc, e.job
    from emp e join dept d
    on (e.deptno = d.deptno)
    where d.loc='DALLAS'

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('JONES', Decimal('2975'), 'DALLAS', 'MANAGER'),
 ('FORD', Decimal('3000'), 'DALLAS', 'ANALYST'),
 ('SMITH', Decimal('800'), 'DALLAS', 'CLERK'),
 ('SCOTT', Decimal('3000'), 'DALLAS', 'ANALYST'),
 ('ADAMS', Decimal('1100'), 'DALLAS', 'CLERK')]

문제 245. emp 테이블과 salgrade 테이블을 서로 조인해서 이름, 월급, 급여등급(grade)를 출력하는데
             2등급만 출력되게 하고 on 절을 사용한 조인문법으로 수행하시오

In [44]:
%%sql

select e.ename, e.sal, s.grade
    from emp e join salgrade s
    on (e.sal between s.losal and s.hisal)
    where s.grade = 2

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('MARTIN', Decimal('1250'), 2),
 ('WARD', Decimal('1250'), 2),
 ('MILLER', Decimal('1300'), 2)]

# 여러 테이블의 데이터를 조인해서 출력하기 5(USING절)
※ using 다음에 공통적으로 들어가는 연결고리가 되는 컬럼명 테이블을 별칭없이 사용

In [45]:
%%sql

select e.ename, d.loc
    from emp e join dept d
    using (deptno)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'NEW YORK')]

문제 246. 직업이 SALESMAN인 사원들의 이름, 월급, 직업, 부서위치를 출력하는데
             using 절을 사용한 join문법으로 수행하시오

In [46]:
%%sql

select e.ename, e.sal, e.job, d.loc
    from emp e join dept d
    using (deptno)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', Decimal('5000'), 'PRESIDENT', 'NEW YORK'),
 ('BLAKE', Decimal('2850'), 'MANAGER', 'CHICAGO'),
 ('CLARK', Decimal('2450'), 'MANAGER', 'NEW YORK'),
 ('JONES', Decimal('2975'), 'MANAGER', 'DALLAS'),
 ('MARTIN', Decimal('1250'), 'SALESMAN', 'CHICAGO'),
 ('ALLEN', Decimal('1600'), 'SALESMAN', 'CHICAGO'),
 ('TURNER', Decimal('1500'), 'SALESMAN', 'CHICAGO'),
 ('JAMES', Decimal('950'), 'CLERK', 'CHICAGO'),
 ('WARD', Decimal('1250'), 'SALESMAN', 'CHICAGO'),
 ('FORD', Decimal('3000'), 'ANALYST', 'DALLAS'),
 ('SMITH', Decimal('800'), 'CLERK', 'DALLAS'),
 ('SCOTT', Decimal('3000'), 'ANALYST', 'DALLAS'),
 ('ADAMS', Decimal('1100'), 'CLERK', 'DALLAS'),
 ('MILLER', Decimal('1300'), 'CLERK', 'NEW YORK')]

# 여러 테이블의 데이터를 조인해서 출력하기 6(NATURAL JOIN)
※ where 절 없이 간단하게 join하는 join문법
오라클이 알아서 두 테이블 사이에 공통된 컬럼이 있는지 찾아보고 join한다.  

<br>

1. 오라클 join문법: 4가지 전부 잘 알고 있어야 함  
2. 1999 ANSI 문법: on 절을 사용한 join 문법과 outer join문법을 잘 알고 있어야 함

In [47]:
%%sql

select e.ename, d.loc
    from emp e natural join dept d

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'NEW YORK')]

# 여러 테이블의 데이터를 조인해서 출력하기 7(LEFT/RIGHT OUTER JOIN)

In [48]:
%%sql

select e.ename, d.loc
    from emp e, dept d
    where e.deptno(+)=d.deptno

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'NEW YORK'),
 (None, 'BOSTON')]

In [49]:
%%sql

select e.ename, d.loc
    from emp e right outer join dept d
    on (d.deptno = e.deptno)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'NEW YORK'),
 (None, 'BOSTON')]

In [50]:
%%sql

select e.ename, d.loc
    from emp e left join dept d
    on (e.deptno = d.deptno)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('KING', 'NEW YORK'),
 ('CLARK', 'NEW YORK'),
 ('MILLER', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('BLAKE', 'CHICAGO'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('jack', None)]

# 여러 테이블의 데이터를 조인해서 출력하기 8(FULL OUTER JOIN)

In [51]:
%%sql

select e.ename, d.loc
    from emp e full outer join dept d
    on (e.deptno = d.deptno)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('jack', None),
 ('KING', 'NEW YORK'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'NEW YORK'),
 (None, 'BOSTON')]

# 여러 테이블의 데이터를 조인해서 출력하기 9(CROSS JOIN)
" where 절 없이 join해서 전체를 다 join하는 문법 "
1. 오라클 문법

In [52]:
%%sql

select e.ename, d.loc
    from emp e, dept d

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('jack', 'NEW YORK'),
 ('KING', 'NEW YORK'),
 ('BLAKE', 'NEW YORK'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'NEW YORK'),
 ('MARTIN', 'NEW YORK'),
 ('ALLEN', 'NEW YORK'),
 ('TURNER', 'NEW YORK'),
 ('JAMES', 'NEW YORK'),
 ('WARD', 'NEW YORK'),
 ('FORD', 'NEW YORK'),
 ('SMITH', 'NEW YORK'),
 ('SCOTT', 'NEW YORK'),
 ('ADAMS', 'NEW YORK'),
 ('MILLER', 'NEW YORK'),
 ('jack', 'DALLAS'),
 ('KING', 'DALLAS'),
 ('BLAKE', 'DALLAS'),
 ('CLARK', 'DALLAS'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'DALLAS'),
 ('ALLEN', 'DALLAS'),
 ('TURNER', 'DALLAS'),
 ('JAMES', 'DALLAS'),
 ('WARD', 'DALLAS'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'DALLAS'),
 ('jack', 'CHICAGO'),
 ('KING', 'CHICAGO'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'CHICAGO'),
 ('JONES', 'CHICAGO'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'CHICAGO'),
 ('SMITH', 'CHICAGO'),
 ('SCOTT', 'CHICAGO'),
 ('ADAMS', 'CHICAGO'),
 ('MILLER', 'CHICAGO'),
 ('jack', 'BOSTON'),
 ('KING', 'BOSTON'),
 ('BLAKE', 'BOSTON'),
 ('CLARK', 'BOSTON'),
 ('JONES', 'BOSTON'),
 ('MARTIN', 'BOSTON'),
 ('ALLEN', 'BOSTON'),
 ('TURNER', 'BOSTON'),
 ('JAMES', 'BOSTON'),
 ('WARD', 'BOSTON'),
 ('FORD', 'BOSTON'),
 ('SMITH', 'BOSTON'),
 ('SCOTT', 'BOSTON'),
 ('ADAMS', 'BOSTON'),
 ('MILLER', 'BOSTON')]

In [53]:
%%sql

select e.ename, d.loc
    from emp e cross join dept d

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('jack', 'NEW YORK'),
 ('KING', 'NEW YORK'),
 ('BLAKE', 'NEW YORK'),
 ('CLARK', 'NEW YORK'),
 ('JONES', 'NEW YORK'),
 ('MARTIN', 'NEW YORK'),
 ('ALLEN', 'NEW YORK'),
 ('TURNER', 'NEW YORK'),
 ('JAMES', 'NEW YORK'),
 ('WARD', 'NEW YORK'),
 ('FORD', 'NEW YORK'),
 ('SMITH', 'NEW YORK'),
 ('SCOTT', 'NEW YORK'),
 ('ADAMS', 'NEW YORK'),
 ('MILLER', 'NEW YORK'),
 ('jack', 'DALLAS'),
 ('KING', 'DALLAS'),
 ('BLAKE', 'DALLAS'),
 ('CLARK', 'DALLAS'),
 ('JONES', 'DALLAS'),
 ('MARTIN', 'DALLAS'),
 ('ALLEN', 'DALLAS'),
 ('TURNER', 'DALLAS'),
 ('JAMES', 'DALLAS'),
 ('WARD', 'DALLAS'),
 ('FORD', 'DALLAS'),
 ('SMITH', 'DALLAS'),
 ('SCOTT', 'DALLAS'),
 ('ADAMS', 'DALLAS'),
 ('MILLER', 'DALLAS'),
 ('jack', 'CHICAGO'),
 ('KING', 'CHICAGO'),
 ('BLAKE', 'CHICAGO'),
 ('CLARK', 'CHICAGO'),
 ('JONES', 'CHICAGO'),
 ('MARTIN', 'CHICAGO'),
 ('ALLEN', 'CHICAGO'),
 ('TURNER', 'CHICAGO'),
 ('JAMES', 'CHICAGO'),
 ('WARD', 'CHICAGO'),
 ('FORD', 'CHICAGO'),
 ('SMITH', 'CHICAGO'),
 ('SCOTT', 'CHICAGO'),
 ('ADAMS', 'CHICAGO'),
 ('MILLER', 'CHICAGO'),
 ('jack', 'BOSTON'),
 ('KING', 'BOSTON'),
 ('BLAKE', 'BOSTON'),
 ('CLARK', 'BOSTON'),
 ('JONES', 'BOSTON'),
 ('MARTIN', 'BOSTON'),
 ('ALLEN', 'BOSTON'),
 ('TURNER', 'BOSTON'),
 ('JAMES', 'BOSTON'),
 ('WARD', 'BOSTON'),
 ('FORD', 'BOSTON'),
 ('SMITH', 'BOSTON'),
 ('SCOTT', 'BOSTON'),
 ('ADAMS', 'BOSTON'),
 ('MILLER', 'BOSTON')]

문제 250. 부서위치, 부서위치별 토탈월급을 가로로 출력하시오 

In [54]:
%%sql

select sum(decode(d.loc, 'NEW YORK', e.sal)) as "NEW YORK",
       sum(decode(d.loc, 'DALLAS', e.sal)) as "DALLAS",
       sum(decode(d.loc, 'CHICAGO', e.sal)) as "CHICAGO"
    from dept d join emp e
    on (d.deptno = e.deptno)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(8750, 10875, 9400)]

# 집합 연산자로 데이터를 위아래로 연결하기 1(UNION ALL)
" 데이터를 연결해서 출력하는 방법 2가지 "  
1. 조인(join): 데이터를 양 옆으로 연결해서 출력하는 방법  
2. 집합 연산자: 데이터를 위 아래로 연결해서 출력하는 방법  


※ 집합 연산자의 종류 4가지
1. union all  
2. union  
3. intersect  
4. minus  

Ex) 직업, 직업별 토탈월급을 출력하시오 ! (세로출력)

In [55]:
%%sql

select job, sum(sal)
    from emp
    group by job

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SALESMAN', 9100),
 ('CLERK', 4150),
 ('ANALYST', 6000),
 ('MANAGER', 8275),
 ('PRESIDENT', 5000)]

Ex) 전체 토탈월급을 출력하시오!

In [62]:
%%sql

select '전체 총 월급:' as job, sum(sal)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('전체 총 월급:', 32525)]

Ex) 위의 SQL을 하나로 합쳐서 데이터가 위아래로 출력되게 하시오

In [63]:
%%sql

select job, sum(sal)
    from emp
    group by job
    union all
select '전체 총 월급:' as job, sum(sal)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SALESMAN', 9100),
 ('CLERK', 4150),
 ('ANALYST', 6000),
 ('MANAGER', 8275),
 ('PRESIDENT', 5000),
 ('전체 총 월급:', 32525)]

※ union all 로 위의 쿼리문의 결과와 아래의 쿼리문의 결과를 하나로 합쳐서 출력  
- 데이터 타입(유형) 3가지  
    1. 문자형  
    2. 숫자형  
    3. 날짜형  

<br>
    
- 집합 연산자를 사용할 때 주의사항은 다음과 같음
    1. 집합 연산자 위 아래의 쿼리문의 컬럼 갯수가 동일해야 함
    2. 집합 연산자 위 아래의 쿼리문의 컬럼 데이터 타입도 동일해야 함
        - 문자형인지 숫자형인지 날짜형인지
    3. 집합연산자 위 아래의 쿼리문의 컬럼명이 동일해야 함
    4. order by 절은 맨 아래 쿼리문에만 작성할 수 있음
        - 정렬할 때는 별칭을 써주고 위 아래의 쿼리문 별칭을 맞춰야 함

문제 251. 부서번호, 부서번호별 토탈월급을 출력하는데 맨 아래에 
             전체 토탈월급이 출력되게 하시오(union all로 수행)

In [81]:
%%sql

select to_char(deptno), sum(sal)
    from emp
    group by deptno
    union all
select '총 합계:', sum(sal)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('70', 3500), ('30', 9400), ('10', 8750), ('20', 10875), ('총 합계:', 32525)]

문제 252. 입사한 년도(4자리), 입사한 년도별 토탈월급을 출력하시오

In [88]:
%%sql

select to_char(hiredate, 'YYYY'), sum(sal)
    from emp
    where to_char(hiredate, 'YYYY') is not null
    group by to_char(hiredate, 'YYYY')

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('1981', 22825), ('1983', 1100), ('1980', 800), ('1982', 4300)]


문제 253. 입사한 년도(4자리), 입사한 년도별 토탈 월급을 출력하는데
             맨 아래에 전체 토탈 월급이 출력되게 하시오!

In [93]:
%%sql

select to_char(hiredate, 'YYYY'), sum(sal)
    from emp
    where to_char(hiredate, 'YYYY') is not null
    group by to_char(hiredate, 'YYYY')
    union all
select '총 합계', sum(sal)
    from emp
    where hiredate is not null

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('1981', 22825),
 ('1983', 1100),
 ('1980', 800),
 ('1982', 4300),
 ('총 합계', 29025)]

In [110]:
%%sql

select nvl(to_char(hiredate, 'YYYY'), '총 합계'), sum(sal)
    from emp
    where to_char(hiredate, 'YYYY') is not null
    group by rollup(to_char(hiredate, 'YYYY'))

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('1980', 800),
 ('1981', 22825),
 ('1982', 4300),
 ('1983', 1100),
 ('총 합계', 29025)]

※ union all 위 아래로 쿼리문의 컬럼의 갯수가 서로 동일해야 하고  
컬럼명도 가급적 동일해야 하고 데이터 유형도 동일해야 함  
문제 254. 우리반 테이블에서 통신사, 통신사별 인원수를 출력하시오

In [97]:
%%sql

select telecom 통신사, count(ename) as 인원수
    from emp12
    group by telecom

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('sk', 10), ('lg', 4), ('kt', 14)]


문제 255. 전체인원수가 맨 아래에 출력 되게 하시오

In [106]:
%%sql

select telecom 통신사, count(ename) 인원수
    from emp12
    group by telecom
    union all
select '총 인원수:' 통신사, count(*) 인원수
    from emp12

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('sk', 10), ('lg', 4), ('kt', 14), ('총 인원수:', 28)]

In [107]:
%%sql

select nvl(telecom, '총 인원수:') 통신사, count(ename) as 인원수
    from emp12
    group by rollup(telecom)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('kt', 14), ('lg', 4), ('sk', 10), ('총 인원수:', 28)]

In [109]:
%%sql

select case when telecom is null then '총 인원수:'
            else telecom end 통신사, 
       count(ename) as 인원수
    from emp12
    group by grouping sets((), telecom)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('kt', 14), ('lg', 4), ('sk', 10), ('총 인원수:', 28)]

문제 256. 입사한 년도(4자리), 입사한 년도별 토탈 월급을 출력하는데 맨 아래에 전체 토탈 월급이 출력되게 하시오!<br>
그리고 이 결과를 입사년도별로 정렬해서 출력하시오

In [120]:
%%sql

select to_char(hiredate, 'YYYY') as "입사년도", sum(sal) as "전체 토탈 월급"
    from emp
    group by to_char(hiredate, 'YYYY')
    union all
select '총 월급' as "입사년도", sum(sal)
    from emp
    where hiredate is not null
    order by 입사년도 asc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('1980', 800),
 ('1981', 22825),
 ('1982', 4300),
 ('1983', 1100),
 ('총 월급', 29025)]

# 집합 연산자로 데이터를 위아래로 연결하기 2(UNION)
" union은 union all과 같은 합집합 연산자인데 차이점은 union은 order by 절을 사용하지 않아도 암시적으로 수행함.  
&nbsp;&nbsp;그리고 중복된 데이터를 하나로 정렬함."

In [121]:
%%sql

select to_char(hiredate, 'RRRR') as hire_year, sum(sal)
    from emp
    group by to_char(hiredate, 'RRRR')
    union
select '토탈:', sum(sal)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('1980', 800),
 ('1981', 22825),
 ('1982', 4300),
 ('1983', 1100),
 ('토탈:', 29025)]

문제 257. 부서번호, 부서번호별 토탈 월급을 출력하는데 맨 아래에  
             전체 토탈 월급이 출력되게 하고 부서번호를 10, 20, 30번 순으로 정렬해서  
             출력되게 하시오

In [125]:
%%sql

select to_char(deptno) as 부서번호, sum(sal) as 총월급
    from emp
    group by deptno
    union
select '총 계:' as 부서번호,sum(sal) as 총월급
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('10', 8750), ('20', 10875), ('30', 9400), ('총 계:', 29025)]

※ union은 암시적으로 정렬 작업을 수행하기 때문에  
굳이 정렬된 결과를 볼 필요가 없다면 union all을 사용하는 게 더 검색성능에 좋음  

<br>
문제 258. 부서위치, 부서위치별 토탈 월급을 출력하시오

In [127]:
%%sql

select d.loc, sum(e.sal)
    from dept d, emp e
    where d.deptno = e.deptno
    group by d.loc

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 8750), ('DALLAS', 10875), ('CHICAGO', 9400)]

문제 259. 맨 아래에 전체 토탈월급도 출력되게 하시오

In [133]:
%%sql

select d.loc "부서위치", sum(e.sal) "월급"
    from dept d join emp e
    on (d.deptno = e.deptno)
    group by d.loc
    union all
select '총 계:' "부서위치", sum(sal) "월급"
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('NEW YORK', 8750), ('DALLAS', 10875), ('CHICAGO', 9400), ('총 계:', 29025)]

In [137]:
%%sql

select nvl(d.loc, '총 계:') "부서위치", sum(e.sal) "월급"
    from dept d natural join emp e
    group by rollup(d.loc)

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('CHICAGO', 9400), ('DALLAS', 10875), ('NEW YORK', 8750), ('총 계:', 29025)]

문제 260. 위의 결과를 다시 출력하는데 직업을 ABCD 순서대로 정렬해서 
             출력되게 하시오

In [138]:
%%sql

select d.loc "부서위치", sum(e.sal) "월급"
    from dept d, emp e
    where d.deptno = e.deptno
    group by d.loc
    union
select '총 계:' as "부서위치", sum(sal) "월급"
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('CHICAGO', 9400), ('DALLAS', 10875), ('NEW YORK', 8750), ('총 계:', 29025)]

※ 정렬을 할 필요가 없으면 union all 을 사용하고 정렬이 필요하면 union 을 사용함


문제 261. 직업, 직업별 최대월급, 직업별 최소월급, 
             직업별 평균얼급, 직업별 인원수를 출력하시오

In [146]:
%%sql

select job, max(sal), min(sal), round(avg(sal)), count(ename)
    from emp
    group by job

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('SALESMAN', 1600, 1250, 1400, 4),
 ('CLERK', 1300, 800, 1038, 4),
 ('ANALYST', 3000, 3000, 3000, 2),
 ('MANAGER', 2975, 2450, 2758, 3),
 ('PRESIDENT', 5000, 5000, 5000, 1)]

문제 262. 사원테이블에서 최대월급, 최소월급, 평균월급, 전체 인원수를 출력하시오

In [148]:
%%sql

select max(sal), min(sal), round(avg(sal)), count(ename)
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[(5000, 800, 2073, 14)]

문제 263. 문제 261번의 결과와 문제 262번의 결과를 위아래로 연결해서 출력하시오

In [152]:
%%sql

select job "직업", max(sal) "최대월급", min(sal) "최소월급", round(avg(sal))"평균월급", count(ename) "총 인원수"
    from emp
    group by job
    union
select '전체 인원 중:' as "직업", max(sal) "최대월급", min(sal) "최소월급", round(avg(sal)) "평균월급", count(ename) "총 인원수"
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('ANALYST', 3000, 3000, 3000, 2),
 ('CLERK', 1300, 800, 1038, 4),
 ('MANAGER', 2975, 2450, 2758, 3),
 ('PRESIDENT', 5000, 5000, 5000, 1),
 ('SALESMAN', 1600, 1250, 1400, 4),
 ('전체 인원 중:', 5000, 800, 2073, 14)]

문제 264. 위의 결과에서 숫자를 전부 천단위라 부여되게 하시오

In [156]:
%%sql

select job "직업", 
       to_char(max(sal), '9,999') "최대월급", 
       to_char(min(sal), '9,999') "최소월급", 
       to_char(round(avg(sal)), '9,999') "평균월급",
       count(ename) "총 인원수"
    from emp
    group by job
    union
select '전체 인원 중:' as "직업", 
        to_char(max(sal), '9,999') "최대월급", 
        to_char(min(sal), '9,999') "최소월급", 
        to_char(round(avg(sal)), '9,999') "평균월급", 
        count(ename) "총 인원수"
    from emp

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('ANALYST', ' 3,000', ' 3,000', ' 3,000', 2),
 ('CLERK', ' 1,300', '   800', ' 1,038', 4),
 ('MANAGER', ' 2,975', ' 2,450', ' 2,758', 3),
 ('PRESIDENT', ' 5,000', ' 5,000', ' 5,000', 1),
 ('SALESMAN', ' 1,600', ' 1,250', ' 1,400', 4),
 ('전체 인원 중:', ' 5,000', '   800', ' 2,073', 14)]

# 집합 연산자로 데이터의 교집합을 출력하기(INTERSECT)
- 집합연산자
    1. 합집합 연산자: union all, union
    2. 교집합 연산자: intersect
    3. 차집합 연산자: minus

※ emp12_backup2 테이블은 김씨 성만을 담은 테이블  

In [219]:
%%sql

select replace(ename, substr(ename, 2, 1), '*') as 이름,
       '*'||substr(age,2, 1) as 나이,
      replace(major, 
              substr(major,1, case when major like '%학과' then regexp_instr(major, '학과', 1)-1 
                                    when major like '%과' then regexp_instr(major, '과', 1)-1 end)
                                                                                                              , '*') as 학과,
      telecom as 통신사,
      substr(email, instr(email, '@')+1) as 이메일, 
      substr(address, 1, 3) as 거주지
    from emp12
    
    intersect

select replace(ename, substr(ename, 2, 1), '*') as 이름,
       '*'||substr(age,2, 1) as 나이,
      replace(major, 
              substr(major,1, case when major like '%학과' then regexp_instr(major, '학과', 1)-1 
                                    when major like '%과' then regexp_instr(major, '과', 1)-1 end)
                                                                                                              , '*') as 학과,
      telecom as 통신사,
      substr(email, instr(email, '@')+1) as 이메일, 
      substr(address, 1, 3) as 거주지
    from emp12_backup2

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('김*라', '*9', '*학과', 'kt', 'naver.com', '서울시'),
 ('김*린', '*7', '*학과', 'kt', 'naver.com', '경기도'),
 ('김*민', '*8', '*과', 'lg', 'naver.com', '경기도'),
 ('김*비', '*4', '*학과', 'kt', 'naver.com', '경기도'),
 ('김*순', '*3', '*학과', 'sk', 'gmail.com', '서울시'),
 ('김*승', '*7', '*학과', 'lg', 'gmail.com', '경기도'),
 ('김*원', '*4', '*학과', 'lg', 'gmail.com', '서울시'),
 ('김*원', '*8', '*학과', 'kt', 'gmail.com', '경기도')]

# 070 집합 연산자로 데이터의 차이를 출력하기(MINUS)  
"차집합을 구하는 집합 연산자"

In [220]:
%%sql

select replace(ename, substr(ename, 2, 1), '*') as 이름,
       '*'||substr(age,2, 1) as 나이,
      replace(major, 
              substr(major,1, case when major like '%학과' then regexp_instr(major, '학과', 1)-1 
                                    when major like '%과' then regexp_instr(major, '과', 1)-1 end)
                                                                                                              , '*') as 학과,
      telecom as 통신사,
      substr(email, instr(email, '@')+1) as 이메일, 
      substr(address, 1, 3) as 거주지
    from emp12
    
    minus

select replace(ename, substr(ename, 2, 1), '*') as 이름,
       '*'||substr(age,2, 1) as 나이,
      replace(major, 
              substr(major,1, case when major like '%학과' then regexp_instr(major, '학과', 1)-1 
                                    when major like '%과' then regexp_instr(major, '과', 1)-1 end)
                                                                                                              , '*') as 학과,
      telecom as 통신사,
      substr(email, instr(email, '@')+1) as 이메일, 
      substr(address, 1, 3) as 거주지
    from emp12_backup2

 * oracle://scott:***@localhost:1521/orcl
0 rows affected.


[('구*모', '*6', '*학과', 'sk', 'naver.com', '서울시'),
 ('권*원', '*6', '*학과', 'sk', 'gmail.com', '서울시'),
 ('권*환', '*8', '*학과', 'kt', 'naver.com', '서울시'),
 ('남*솔미', '*7', '*학과', 'kt', 'yahoo.com', '경기도'),
 ('성*창', '*9', '*학과', 'sk', 'gmail.com', '서울시'),
 ('송*미', '*9', '*학과', 'kt', 'naver.com', '경기도'),
 ('신*종', '*7', '*학과', 'lg', 'naver.com', '경기도'),
 ('양*준', '*6', '*학과', 'sk', 'naver.com', '서울시'),
 ('유*영', '*8', '*학과', 'sk', 'naver.com', '경기도'),
 ('이*성', '*7', '*학과', 'sk', 'naver,com', '강원도'),
 ('이*원', '*5', '*학과', 'kt', 'gmail.com', '경기도'),
 ('이*혁', '*9', '*학과', 'kt', 'naver.com', '경기도'),
 ('장*진', '*5', '*학과', 'kt', 'naver.com', '경기도'),
 ('정*원', '*4', '*학과', 'sk', 'naver.com', '경기도'),
 ('정*희', '*6', '*학과', 'kt', 'naver.com', '서울시'),
 ('한*', '*1', '*학과', 'kt', 'naver.com', '경기도'),
 ('허*', '*6', '*학과', 'kt', 'naver.com', ' 경기'),
 ('허*민', '*0', '*학과', 'sk', 'hanmail.net', '경기도'),
 ('현*연', '*5', '*학과', 'sk', 'naver.com', '경기도'),
 ('황*현', '*7', '*학과', 'kt', 'naver.com', '서울 ')]